expression/extension.go

```diff
@@ -0,0 +1,185 @@
+
+func registerExtensionFunc(def *extension.FunctionDef) error {
+	if def == nil {
+		return errors.New("extension function def is nil")
+	}
+
+	if def.Name == "" {
+		return errors.New("extension function name should not be empty")
+	}
+
+	lowerName := strings.ToLower(def.Name)
+	if _, ok := funcs[lowerName]; ok {
+		return errors.Errorf("extension function name '%s' conflict with builtin", def.Name)
+	}
+
+	class, err := newExtensionFuncClass(def)
+	if err != nil {
+		return err
+	}
+
+	_, exist := extensionFuncs.LoadOrStore(lowerName, class)
+	if exist {
+		return errors.Errorf("duplicated extension function name '%s'", def.Name)
+	}
+
+	return nil
+}
+
+func removeExtensionFunc(name string) {
+	extensionFuncs.Delete(name)
+}
+
+type extensionFuncClass struct {
+	baseFunctionClass
+	funcDef *extension.FunctionDef
+	flen    int
+}
+
+func newExtensionFuncClass(def *extension.FunctionDef) (*extensionFuncClass, error) {
+	var flen int
+	switch def.EvalTp {
+	case types.ETString:
+		flen = mysql.MaxFieldVarCharLength
+		if def.EvalStringFunc == nil {
+			return nil, errors.New("eval function is nil")
+		}
+	case types.ETInt:
+		flen = mysql.MaxIntWidth
+		if def.EvalIntFunc == nil {
+			return nil, errors.New("eval function is nil")
+		}
+	default:
+		return nil, errors.Errorf("unsupported extension function ret type: '%v'", def.EvalTp)
+	}
+
+	return &extensionFuncClass{
+		baseFunctionClass: baseFunctionClass{def.Name, len(def.ArgTps), len(def.ArgTps)},
+		flen:              flen,
+		funcDef:           def,
+	}, nil
+}
+
+func (c *extensionFuncClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
+	if err := c.checkPrivileges(ctx); err != nil {
+		return nil, err
+	}
+
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, c.funcDef.EvalTp, c.funcDef.ArgTps...)
+	if err != nil {
+		return nil, err
+	}
+	bf.tp.SetFlen(c.flen)
+	sig := &extensionFuncSig{bf, c.funcDef}
+	return sig, nil
+}
+
+func (c *extensionFuncClass) checkPrivileges(ctx sessionctx.Context) error {
+	privs := c.funcDef.RequireDynamicPrivileges
+	if semPrivs := c.funcDef.SemRequireDynamicPrivileges; len(semPrivs) > 0 && sem.IsEnabled() {
+		privs = semPrivs
+	}
+
+	if len(privs) == 0 {
+		return nil
+	}
+
+	manager := privilege.GetPrivilegeManager(ctx)
+	activeRoles := ctx.GetSessionVars().ActiveRoles
+
+	for _, priv := range privs {
+		if !manager.RequestDynamicVerification(activeRoles, priv, false) {
+			msg := priv
+			if !sem.IsEnabled() {
+				msg = "SUPER or " + msg
+			}
+			return errSpecificAccessDenied.GenWithStackByArgs(msg)
+		}
+	}
+
+	return nil
+}
+
+var _ extension.FunctionContext = &extensionFuncSig{}
+
+type extensionFuncSig struct {
+	baseBuiltinFunc
+	*extension.FunctionDef
+}
+
+func (b *extensionFuncSig) Clone() builtinFunc {
+	newSig := &extensionFuncSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	newSig.FunctionDef = b.FunctionDef
```

Comment:
Is there any chance that `FunctionDef` might be changed? 

I just want to confirm that it should be immutable, if so that's is fine here.

Reply:
Is it immutble in current implement. However, I still make the pointer to builtin object to make it safer